### DISCUSSION TITLE
Fetching normals and tangents of a Tri6

### DISCUSSION THREAD

**Discussion by rebpoli** [libMesh/libmesh#3073] [Q&A] [answered]
Hi,

It seems the get_tangents method only works on faces (sides of a Tet, for example).
I need tangents and normals for a Tri6.
Do libmesh interfaces provide that?

Thanks,
Renato

**Comment by jwpeterson** [comment, marked as answer]
You can call:
```
const std::vector<RealGradient>& dxyzdxi = fe->get_dxyzdxi();
const std::vector<RealGradient>& dxyzdeta = fe->get_dxyzdeta();
```
Then the "face" normal to the Tri6 at quadrature point `qp` should be given by:
```
dxyzdxi[qp].cross(dxyzdeta[qp]).unit()
```

**Comment by rebpoli** [comment]
Thanks,

If I got it properly, I would have:
Point tan1 = dxyzdxi[qp].unit();
Point tan2 = dxyzdxi[qp].unit();
Point normal = dxyzdxi[qp].cross(dxyzdeta[qp]).unit();

Is that right?

Em qua., 3 de nov. de 2021 às 16:01, John W. Peterson <
***@***.***> escreveu:

**Comment by jwpeterson** [comment]
Yes, I think that should work, but it would be good if you could check that the results you get seem reasonable.

**Reply by jwpeterson** [reply]
Actually your "tan2" should use dxyzdeta.

**Comment by rebpoli** [comment]
Sorry for the typo.

Well .. Below is what I got (on the left would be the expected, on the
right the resulting).
I can see that *dxyzdxi* is not orthogonal to *dxyzdeta*.
Is that expected?

Point tan_s    = dxyzdxi[qp].unit();
Point tan_t    = dxyzdeta[qp].unit();
Point normal_f = dxyzdxi[qp].cross(dxyzdeta[qp]).unit();

tan_s:   (x,y,z)=(      -1,        0,        0)
tan_t:   (x,y,z)=(-0.672813,       0, 0.739813)
normal_f:(x,y,z)=(       0,        1,        0)


Em qua., 3 de nov. de 2021 às 16:14, John W. Peterson <
***@***.***> escreveu:

**Comment by jwpeterson** [comment]
OK, I have actually only used this formula with Quad4 and Tri3 and it seemed to work OK for this, but maybe it doesn't generalize? I would suggest you find the orthogonal projection of tan_t onto tan_s in the plane spanned by tan_s and tan_t and use that for the second tangent vector instead if it's important for the two tangent directions to be orthogonal...

**Comment by rebpoli** [comment]
Thanks!

Em qua., 3 de nov. de 2021 às 16:36, John W. Peterson <
***@***.***> escreveu:
